services/src/main/java/org/keycloak/services/resources/admin/ComponentResource.java

```diff
@@ -126,19 +129,24 @@ public Stream<ComponentRepresentation> getComponents(@QueryParam("parent") Strin
     @Consumes(MediaType.APPLICATION_JSON)
     public Response create(ComponentRepresentation rep) {
         auth.realm().requireManageRealm();
-        try {
-            ComponentModel model = RepresentationToModel.toModel(session, rep);
-            if (model.getParentId() == null) model.setParentId(realm.getId());
+        AtomicReference<Response> ref = new AtomicReference<>();
+        KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), kcSession -> {
+            RealmModel realmModel = LockObjectsForModification.lockRealmsForModification(kcSession, () -> kcSession.realms().getRealm(realm.getId()));
```

Comment:
> Some time ago we reasoned that `JpaComponentEntity` should be a `JpaRootVersionedEntity`, thereby allowing that one component can be modified independently from other components in a realm. We said they are independent (and that's different from mere attributes).
> 
> Applying that reasoning to this code, one would lock only a single component, not the whole realm.

The problem, I think, is that this independence is limited in the sense that components are manipulated through the realm - we don't have a provider/store/transaction for components like we have for e.g. clients, roles, etc, and adding all this structure just to satisfy a concurrent component creation on a specific DB seems completely out of bounds to me.

The `RETRY_SERIALIZABLE` happens upon concurrent creation of components and I think this has to do with the fact that they are added to the mapped collection in the realm and somehow concurrently messing with this collection is causing issues on CRDB, so I had to go for the realm locking. Also, how would I lock a component that still doesn't exist in the DB?

Reply:
Yes, getting the whole list of components first prevents a serializable transaction later when adding an entry. This independence only applies for READ_COMMITTED, and not SERIALIZABLE.

Retrieving the full list is triggered in several places: 

* on the physical layer: the collection we have in JPA that is always loaded in full, even if we would access one component by ID.
* on the service layer: when retrieving `DefaultKeyManager.getProviders` it calls `realm.getComponentsStream(realm.getId(), KeyProvider.class.getName())` which will iterate over all providers in the collection again. 

To make this more efficient, one would need to get rid of the JPA collection and remodel the MapRealmEntity to offer not only getComponents(). Although it looks like an optimization, it is affecting interfaces on all map implementations. I see that we have all necessary columns in the JPA table already, although they are not used AFAIK. A new issue for this (not so small) optimization?

Reply:
> Some time ago we reasoned that `JpaComponentEntity` should be a `JpaRootVersionedEntity`, thereby allowing that one component can be modified independently from other components in a realm. We said they are independent (and that's different from mere attributes).

The component is independent in the sense of where its data is stored (in a separate table). Yet the respective `RealmModel` does change upon a change of a component. In other words, the respective `MapRealmEntity` contents has changed as well upon change of a component within that realm, even if it is not directly a change in the contents of the appropriate `JpaRealmEntity`. This has to be visible to all nodes in the cluster, so that they could update the instantiated component based on the updated definition. Is this satisfied? Cc @vramik

Reply:
The JPA store has the concept of a `JpaChildEntity`, which would trigger a Hibernate version counter increment on the parent JPA entity. In a previous discussion we came to the conclusion that the two would be independent, and therefore `JpaComponentEntity` doesn't implement this. This decision was also documented in the JavaDoc of the class. 

https://github.com/keycloak/keycloak/blob/c31d37ddf1c3e3529e761800a6842fdbf25ba931/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/realm/entity/JpaComponentEntity.java#L50-L54

If this is not the case, then this needs to be changed. 

On the other hand, the Hibernate version increments is a concept that is only known to the JPA store, other stores doesn't implement something like this AFAIK.

You wrote: 

> This has to be visible to all nodes in the cluster, so that they could update the instantiated component based on the updated definition.

I'm not aware that there is an interface/mechanism on the Map storage interface level to support notifications  components to be re-instantiated.

Reply:
> If this is not the case, then this needs to be changed.

I don't think the behaviour descibed in Javadoc reflects the expected behaviour.

> I'm not aware that there is an interface/mechanism on the Map storage interface level to support notifications components to be re-instantiated.

Invalidation in `KeycloakSessionFactory`.

For example, in case of detected realm update by JPA provider, JPA provider should invalidate the realm via `session.invalidate(ObjectType.REALM, realmId)` which in turn invalidates the respective components.

There is no mechanism similar to `ClusterProvider` from the legacy store neither implemented nor anticipated. Each implementation provides its own way of invalidation detection appropriate for the respective technology; in case of JPA, that might be detecting change in object version of an entity.

Reply:
@hmlnarik - looking into `MapRealmAdapter.updateComponent` there is already a call to `ComponentUtil.notifyUpdated` that will via the `AmphibianProviderFactory.onUpdate` will call `session.invalidate(ObjectType.COMPONENT, ...` if I understand this correctly.

If such a notification by component is not sufficient, please raise an issue for the invalidation of the realm. If such an invalidation would need to be implemented, it would be great to implement this on the level of `MapRealmAdapter` so that it is independent of the underlying physical store.